File: lr6_z1.py
import numpy as np
import matplotlib.pyplot as plt

# ---------------------------------------------------------------------------
# ЗАДАНИЕ:
# Уравнение: x*(2*x-1)*y' + y^2 - (4*x+1)*y + 4*x = 0
# Начальное условие: y(0) = 1
# Интервал: x от 0 до 1
# Шаг: h = 0.1
#
# Для использования численных методов, выразим y' (производную):
# y' = (-y^2 + (4*x+1)*y - 4*x) / (x*(2*x-1))
# ---------------------------------------------------------------------------

def f(x, y):
    """
    Правая часть дифференциального уравнения y' = f(x, y).
    """
    # ОСОБЕННОСТЬ 1: Устранимый разрыв в x=0.
    # В точке x=0 знаменатель обращается в ноль. Предел y' при x -> 0 равен -1.
    if np.isclose(x, 0.0):
        return -1.0
    
    # ОСОБЕННОСТЬ 2: Устранимый разрыв в x=0.5.
    # В точке x=0.5 знаменатель также обращается в ноль.
    # Можно показать (например, с помощью правила Лопиталя), что предел
    # y' при x -> 0.5 для точного решения равен 2/3.
    if np.isclose(x, 0.5):
        return 2.0 / 3.0

    denominator = x * (2 * x - 1)
    numerator = -y**2 + (4 * x + 1) * y - 4 * x
    return numerator / denominator

def exact_solution(x):
    """
    Точное решение задачи Коши для сравнения.
    y(x) = (2*x^2 + 1) / (x + 1)
    Эта функция теперь векторизована и работает с массивами NumPy.
    """
    return (2 * np.power(x, 2) + 1) / (x + 1)

# --- Параметры задачи ---
x0 = 0.0
y0 = 1.0
h = 0.1
x_end = 1.0

# Создаем массив точек x, на которых будем искать решение
x_points = np.arange(x0, x_end + h, h)
n_points = len(x_points)

# --- Инициализация массивов для хранения решений ---
y_euler = np.zeros(n_points)
y_rk4 = np.zeros(n_points)
y_adams = np.zeros(n_points)

# Задаем начальные условия
y_euler[0] = y0
y_rk4[0] = y0
y_adams[0] = y0

# --- 1. Метод Эйлера ---
for i in range(n_points - 1):
    y_euler[i+1] = y_euler[i] + h * f(x_points[i], y_euler[i])

# --- 2. Метод Рунге-Кутта 4-го порядка ---
for i in range(n_points - 1):
    xi = x_points[i]
    yi = y_rk4[i]
    
    k1 = h * f(xi, yi)
    k2 = h * f(xi + 0.5 * h, yi + 0.5 * k1)
    k3 = h * f(xi + 0.5 * h, yi + 0.5 * k2)
    k4 = h * f(xi + h, yi + k3)
    
    y_rk4[i+1] = yi + (k1 + 2*k2 + 2*k3 + k4) / 6


y_exact = exact_solution(x_points)

# --- Вывод результатов в консоль в виде таблицы ---
print("="*65)
print(f"{'x':>5} | {'Точное y(x)':>15} | {'Метод Эйлера':>15} | {'Метод Рунге-Кутта':>20}")
print("-"*65)
for i in range(n_points):
    # Используем np.nan_to_num, чтобы заменить возможные nan/inf на 0 для вывода,
    # хотя после исправлений их быть не должно.
    print(f"{x_points[i]:>5.1f} | {y_exact[i]:>15.4f} | {np.nan_to_num(y_euler[i]):>15.4f} | {np.nan_to_num(y_rk4[i]):>20.4f} ")
print("="*65)


# --- Построение графиков ---
plt.style.use('seaborn-v0_8-whitegrid')
plt.figure(figsize=(12, 8))

plt.plot(x_points, y_exact, 'o-', label='Точное решение', linewidth=2, zorder=5)
plt.plot(x_points, y_euler, 's--', label='Метод Эйлера', alpha=0.8)
plt.plot(x_points, y_rk4, '^-', label='Метод Рунге-Кутта 4', alpha=0.8)

# Настройка графика
plt.title('Сравнение численных методов решения задачи Коши', fontsize=16)
plt.xlabel('x', fontsize=12)
plt.ylabel('y(x)', fontsize=12)
plt.legend(fontsize=12)
plt.grid(True)
plt.show()

File: lr6_z1.m
% ЗАДАНИЕ:
% Уравнение: x*(2*x-1)*y' + y^2 - (4*x+1)*y + 4*x = 0
% y' = (-y^2 + (4*x+1)*y - 4*x) / (x*(2*x-1))
% Начальное условие: y(0) = 1
% Интервал: x от 0 до 1
% Шаг: h = 0.1

clear;
clc;
close all;

x0 = 0.0;
y0 = 1.0;
h = 0.1;
x_end = 1.0;

% Создаем вектор узлов сетки
x_points = x0:h:x_end;
n_points = length(x_points);

% Инициализация векторов для хранения решений
y_euler = zeros(1, n_points);
y_rk4 = zeros(1, n_points);

% Задаем начальные условия для каждого метода
y_euler(1) = y0;
y_rk4(1) = y0;

% Численное решение

% Метод Эйлера
for i = 1:(n_points - 1)
    x_current = x_points(i);
    y_current = y_euler(i);
    
    % Расчет производной y' = f(x,y)
    if abs(x_current - 0.0) < 1e-9
        dydx = -1.0;
    elseif abs(x_current - 0.5) < 1e-9
        dydx = 2.0 / 3.0;
    else
        num = -y_current^2 + (4*x_current + 1)*y_current - 4*x_current;
        dem = x_current * (2*x_current - 1);
        dydx = num / dem;
    end
    
    y_euler(i+1) = y_current + h * dydx;
end

% Метод Рунге-Кутта 4-го порядка
for i = 1:(n_points - 1)
    xi = x_points(i);
    yi = y_rk4(i);
    
    % Расчет k1
    x_k = xi;
    y_k = yi;
    if abs(x_k - 0.0) < 1e-9
        dydx1 = -1.0;
    elseif abs(x_k - 0.5) < 1e-9
        dydx1 = 2.0 / 3.0;
    else
        num = -y_k^2 + (4*x_k + 1)*y_k - 4*x_k;
        dem = x_k * (2*x_k - 1);
        dydx1 = num / dem;
    end
    k1 = h * dydx1;
    
    % Расчет k2
    x_k = xi + 0.5 * h;
    y_k = yi + 0.5 * k1;
    if abs(x_k - 0.0) < 1e-9
        dydx2 = -1.0;
    elseif abs(x_k - 0.5) < 1e-9
        dydx2 = 2.0 / 3.0;
    else
        num = -y_k^2 + (4*x_k + 1)*y_k - 4*x_k;
        dem = x_k * (2*x_k - 1);
        dydx2 = num / dem;
    end
    k2 = h * dydx2;

    % Расчет k3
    x_k = xi + 0.5 * h;
    y_k = yi + 0.5 * k2;
    if abs(x_k - 0.0) < 1e-9
        dydx3 = -1.0; 
    elseif abs(x_k - 0.5) < 1e-9
        dydx3 = 2.0 / 3.0;
    else
        num = -y_k^2 + (4*x_k + 1)*y_k - 4*x_k;
        dem = x_k * (2*x_k - 1);
        dydx3 = num / dem;
    end
    k3 = h * dydx3;
    
    % Расчет k4
    x_k = xi + h;
    y_k = yi + k3;
    if abs(x_k - 0.0) < 1e-9
        dydx4 = -1.0;
    elseif abs(x_k - 0.5) < 1e-9
        dydx4 = 2.0 / 3.0;
    else
        num = -y_k^2 + (4*x_k + 1)*y_k - 4*x_k;
        dem = x_k * (2*x_k - 1);
        dydx4 = num / dem;
    end
    k4 = h * dydx4;

    % Итоговый шаг
    y_rk4(i+1) = yi + (k1 + 2*k2 + 2*k3 + k4) / 6;
end

%Точное решение и вывод результатов
% функция для точного решения y(x) = (2*x^2 + 1) / (x + 1)
exact_solution = @(x) (2*x.^2 + 1) ./ (x + 1);

% Вычисляем точное решение в узлах сетки
y_exact = exact_solution(x_points);

fprintf('=========================================================================\n');
fprintf('|   x   |   Точное y(x)  |  Метод Эйлера  |  Метод Рунге-Кутта  |\n');
fprintf('=========================================================================\n');
for i = 1:n_points
    fprintf('| %5.1f |    %10.4f  |   %10.4f   |     %10.4f      |\n', ...
        x_points(i), y_exact(i), y_euler(i), y_rk4(i));
end
fprintf('=========================================================================\n');

%Построение графиков
figure('Name', 'Сравнение численных методов', 'NumberTitle', 'off'); 
hold on;

% Строим графики для каждого решения
plot(x_points, y_exact, 'o-', 'LineWidth', 2, 'DisplayName', 'Точное решение');
plot(x_points, y_euler, 's--', 'LineWidth', 1.5, 'DisplayName', 'Метод Эйлера');
plot(x_points, y_rk4, '^-', 'LineWidth', 1.5, 'DisplayName', 'Метод Рунге-Кутта 4');

% Настройка внешнего вида графика
title('Сравнение численных методов решения задачи Коши');
xlabel('x');
ylabel('y(x)');
legend('show', 'Location', 'northwest');
